### PR TITLE
:recycle: (tooltip) refactoring to react-aria (vol.3)

### PR DESCRIPTION
**Pull request by MatissJanis**
Follow up to https://github.com/actualbudget/actual/pull/2493

This refactors the autocompletes, date selects and the "filters" menu.

### CODE REVIEW

**Comment by MatissJanis** [comment]
Out for review now.

The VRT update is unrelated to this PR. VRT is currently failing in `master` branch.

**Comment by joel-jeremy** [comment]
Would it be possible to retain the old scrollbars?

Old (Demo link):
image
image
image

New (This PR):
image
image
image


Observations:
- Fliters: The popovers are not closing when autocomplete popovers are open and you click outside

**Comment by MatissJanis** [comment]
Scrollbar: I pushed a change. Does the scrollbars work as expected now? Totally agree with you that we need to get this patched before merging this PR. But sadly I'm not able to reproduce on OSX, so fixing it is very challenging.

> Fliters: The popovers are not closing when autocomplete popovers are open and you click outside

This one's tricky and I think we need to make a decision here on whether we proceed with the migration to `react-aria` or we roll back the changes. I don't think we'll be able to make it 1:1 match our current experience with `react-aria`.

Here's my investigation:

When we use stacked popovers (filters popover + autocomplete) - [only the topmost popover](https://github.com/adobe/react-spectrum/blob/a597a0ce5f345a7c5968939d120eafbe6f02ac2a/packages/%40react-aria/overlays/src/useOverlay.ts#L87-L92) can be closed when clicking on the outside. This is also assuming none of the popovers use the `isNonModal` prop. So best case scenario: closing the filter popover would require 2x clicks on outside (1x click would close autocomplete and 1x closes the filter menu).

However, we do want to use `isNonModal` for autocompletes. `isNonModal` makes sure than an "underlay" is not added to the page. The underlay intercepts clicks outside the popover context in order to close it. It also kills the target clicks. For example: if you clicked on a button while the underlay was active - the underlay would intercept the button click and thus the button would NOT be clicked.

`isNonModal` is important for transactions table where a user can click on the next row - this should immediately open the next row for editing and close the current.

So to sum it up: we have to decide if we wish to proceed with the migration given the functional change that you discovered. Clicking outside the filters menu while the autocomplete is open does not close the filter menu. WDYT?

**Comment by joel-jeremy** [comment]
The scroll bars are all good now!

> closing the filter popover would require 2x clicks on outside (1x click would close autocomplete and 1x closes the filter menu).

I'm good with this change in behavior. I just noticed something weird when the filter auto complete is open. When I click outside the filter popover, the auto complete popover won't close - it just returns focus to the auto complete input field. Only when click on the sidebar then that's when it closes.

**Comment by youngcw** [comment]
Im still getting the basic scroll bars in Brave on Linux.  Firefox is working though

**Comment by MatissJanis** [comment]
> Im still getting the basic scroll bars in Brave on Linux. Firefox is working though

Could you help me out here and submit a commit for the fix here? Normally I'd be happy to work on it, but I have only OSX :/

**Comment by shall0pass** [comment]
Are the scrollbar's appearance supposed to be unchanged from the current demo?
I'm getting this on Edge (Windows 10):
image

And on Chrome (Windows 10):
image

**Comment by youngcw** [comment]
> Are the scrollbar's appearance supposed to be unchanged from the current demo? I'm getting this on Edge (Windows 10): image
> 
> And on Chrome (Windows 10): image

They are _supposed to_ not change, but are currently stuck trying to figure out what the problem is

**Comment by youngcw** [comment]
Ive been poking around this scrollbar issue.  I get this error in the console.  I tried changing the property name like it suggests, but nothing changed other than the error went away.
<details>

  <summary>Error</summary>

```
Warning: Unsupported style property & ::-webkit-scrollbar. Did you mean & ::WebkitScrollbar?
    at div
    at $9bf71ea28793e738$export$20e40289641fbbb6 (￼http://localhost:3001/node_modules/.vite/deps/chunk-DIVEDOFX.js?v=dac6d417:1657:9)
    at $f1ab8c75478c6f73$export$cf75428e0b9ed1ea (￼http://localhost:3001/node_modules/.vite/deps/chunk-DIVEDOFX.js?v=dac6d417:740:54)
    at $337b884510726a0d$export$c6fdb837b070b4ff (￼http://localhost:3001/node_modules/.vite/deps/react-aria-components.js?v=dac6d417:4591:61)
    at $07b14b47974efb58$var$PopoverInner (￼http://localhost:3001/node_modules/.vite/deps/react-aria-components.js?v=dac6d417:23083:47)
    at $07b14b47974efb58$var$Popover (￼http://localhost:3001/node_modules/.vite/deps/react-aria-components.js?v=dac6d417:23055:64)
    at Popover (￼http://localhost:3001/src/components/common/Popover.tsx?t=1714242015775:12:27)
    at div
    at Downshift3 (￼http://localhost:3001/node_modules/.vite/deps/downshift.js?v=dac6d417:597:26)
    at SingleAutocomplete (￼http://localhost:3001/src/components/autocomplete/Autocomplete.tsx?t=1714242015775:138:31)
    at Autocomplete (￼http://localhost:3001/src/components/autocomplete/Autocomplete.tsx?t=1714242015775:598:35)
    at CategoryAutocomplete (￼http://localhost:3001/src/components/autocomplete/CategoryAutocomplete.tsx?t=1714242015775:116:40)
    at ConditionalPrivacyFilter (￼http://localhost:3001/src/components/PrivacyFilter.tsx:16:44)
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at Cell (￼http://localhost:3001/src/components/table.tsx?t=1714242015775:110:24)
    at CustomCell (￼http://localhost:3001/src/components/table.tsx?t=1714242015775:373:37)
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at Row (￼http://localhost:3001/src/components/table.tsx?t=1714242015775:218:23)
    at Transaction (￼http://localhost:3001/src/components/transactions/TransactionsTable.jsx?t=1714242015775:578:26)
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at div
    at div
    at FixedSizeList (￼http://localhost:3001/src/components/FixedSizeList.tsx:26:9)
    at AvoidRefocusScrollProvider (￼http://localhost:3001/src/hooks/useProperFocus.tsx:40:46)
    at div
    at AutoSizer (￼http://localhost:3001/node_modules/.vite/deps/react-virtualized-auto-sizer.js?v=dac6d417:194:5)
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at ￼http://localhost:3001/src/components/table.tsx?t=1714242015775:756:54
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at TransactionTableInner (￼http://localhost:3001/src/components/transactions/TransactionsTable.jsx?t=1714242015775:1411:34)
    at ￼http://localhost:3001/src/components/transactions/TransactionsTable.jsx?t=1714242015775:1653:51
    at TransactionList (￼http://localhost:3001/src/components/transactions/TransactionList.jsx?t=1714242015775:53:35)
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at SelectedProvider (￼http://localhost:3001/src/hooks/useSelected.tsx:234:36)
    at SelectedProviderWithItems (￼http://localhost:3001/src/hooks/useSelected.tsx:291:45)
    at AllTransactions (￼http://localhost:3001/src/components/accounts/Account.jsx?t=1714242015775:120:28)
    at AccountInternal (￼http://localhost:3001/src/components/accounts/Account.jsx?t=1714242015775:212:9)
    at AccountHack (￼http://localhost:3001/src/components/accounts/Account.jsx?t=1714242015775:1554:50)
    at SplitsExpandedProvider (￼http://localhost:3001/src/hooks/useSplitsExpanded.jsx:25:42)
    at SchedulesProvider (￼http://localhost:3001/@fs/home/caleb/Desktop/actual/packages/loot-core/src/client/data-hooks/schedules.tsx:59:37)
    at Account (￼http://localhost:3001/src/components/accounts/Account.jsx?t=1714242015775:1575:20)
    at LoadComponentInner (￼http://localhost:3001/src/components/util/LoadComponent.tsx?t=1714242015775:33:31)
    at LoadComponent (￼http://localhost:3001/src/components/util/LoadComponent.tsx?t=1714242015775:23:14)
    at NarrowAlternate (￼http://localhost:3001/src/components/responsive/index.tsx?t=1714242015775:26:35)
    at RenderedRoute (￼http://localhost:3001/node_modules/.vite/deps/react-router-dom.js?v=dac6d417:3547:5)
    at Routes (￼http://localhost:3001/node_modules/.vite/deps/react-router-dom.js?v=dac6d417:3982:5)
    at div
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at Router (￼http://localhost:3001/node_modules/.vite/deps/react-router-dom.js?v=dac6d417:3925:15)
    at BrowserRouter (￼http://localhost:3001/node_modules/.vite/deps/react-router-dom.js?v=dac6d417:4660:5)
    at FinancesAppWithoutContext (￼http://localhost:3001/src/components/FinancesApp.tsx?t=1714242015775:125:21)
    at ScrollProvider (￼http://localhost:3001/src/components/ScrollProvider.tsx:15:34)
    at DndProvider2 (￼http://localhost:3001/node_modules/.vite/deps/react-dnd.js?v=dac6d417:1329:9)
    at BudgetMonthCountProvider (￼http://localhost:3001/src/components/budget/BudgetMonthCountContext.tsx:14:44)
    at SidebarProvider (￼http://localhost:3001/src/components/sidebar/SidebarProvider.tsx:16:35)
    at TitlebarProvider (￼http://localhost:3001/src/components/Titlebar.tsx?t=1714242015775:49:36)
    at SpreadsheetProvider (￼http://localhost:3001/@fs/home/caleb/Desktop/actual/packages/loot-core/src/client/SpreadsheetProvider.tsx:118:39)
    at FinancesApp (￼http://localhost:3001/src/components/FinancesApp.tsx?t=1714242015775:521:17)
    at AppInner (￼http://localhost:3001/src/components/App.tsx?t=1714242015775:35:21)
    at ErrorBoundary (￼http://localhost:3001/node_modules/.vite/deps/react-error-boundary.js?v=dac6d417:18:5)
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at div
    at _c (￼http://localhost:3001/src/components/common/View.tsx:17:13)
    at ResponsiveProvider (￼http://localhost:3001/src/ResponsiveProvider.tsx:24:7)
    at BoundHotkeysProxyProviderProvider (￼http://localhost:3001/node_modules/.vite/deps/react-hotkeys-hook.js?v=dac6d417:223:24)
    at HotkeysProvider2 (￼http://localhost:3001/node_modules/.vite/deps/react-hotkeys-hook.js?v=dac6d417:252:36)
    at App (￼http://localhost:3001/src/components/App.tsx?t=1714242015775:158:24)
    at ServerProvider (￼http://localhost:3001/src/components/ServerContext.tsx:41:34)
    at Provider (￼http://localhost:3001/node_modules/.vite/deps/react-redux.js?v=dac6d417:453:20)
```

</details>

**Comment by MatissJanis** [comment]
Pushed a small update for the scrollbars. Let me know how it looks like now on your devices. 🤞

**Comment by youngcw** [comment]
Its definitely better, but still not the same as before
image

**Comment by youngcw** [comment]
Still the same.  Interestingly, it works when in responsive mode, even when viewing the desktop style view.

**Comment by youngcw** [comment]
The scrollbar on the budget page is broken now.  I don't know when that started, or if its happening to anyone else

**Comment by MatissJanis** [comment]
Pushed an update. The scrollbar changes will now apply only to the new popovers. Which should make budget page work just as it did before.

Though I still don't have ideas how to reduce the scrollbar width to match the one we had before.

**Comment by youngcw** [comment]
Ok, that fixed the budget page scrollbars

File: packages/desktop-client/src/components/autocomplete/Autocomplete.tsx [hunk 1]
```diff
@@ -559,17 +561,21 @@ function SingleAutocomplete<T extends Item>({
...
+              <Popover
+                triggerRef={triggerRef}
+                placement="bottom start"
                 offset={2}
+                isOpen={isOpen}
+                onOpenChange={close}
+                isNonModal
                 style={{
-                  padding: 0,
+                  ...styles.darkScrollbar,
```
Comment (joel-jeremy) [inline packages/desktop-client/src/components/autocomplete/Autocomplete.tsx:572]: Instead of putting this in a `style` prop, can we try using glamor to pass this to `className`? Maybe the inline `style` prop doesn't support pseudo selectors.

Comment (MatissJanis) [reply packages/desktop-client/src/components/autocomplete/Autocomplete.tsx:572]: Using glamor is a good idea! Updated it now, let me know how it looks!

Comment (joel-jeremy) [reply packages/desktop-client/src/components/autocomplete/Autocomplete.tsx:572]: It still is not working, we need to check that the darkScrollbar is set in the children's container element.

File: packages/desktop-client/src/components/autocomplete/Autocomplete.tsx [hunk 2]
```diff
@@ -559,17 +561,21 @@ function SingleAutocomplete<T extends Item>({
...
                   backgroundColor: theme.menuAutoCompleteBackground,
                   color: theme.menuAutoCompleteText,
                   minWidth: 200,
-                  ...tooltipStyle,
+                  width: triggerRef.current?.clientWidth,
                 }}
-                {...tooltipProps}
                 data-testid="autocomplete"
               >
                 {renderItems(
```
Comment (joel-jeremy) [inline packages/desktop-client/src/components/autocomplete/Autocomplete.tsx:581]: Maybe we can try wrapping this in a View and setting the scrollbar style there?

Comment (MatissJanis) [reply packages/desktop-client/src/components/autocomplete/Autocomplete.tsx:581]: Would you be open to try this out locally? I'm working blindly here without a way to reproduce.

File: packages/desktop-client/src/components/common/Popover.tsx [hunk 3]
```diff
@@ -1,16 +1,43 @@
...
+  ...props
+}: PopoverProps) => {
   return (
     <ReactAriaPopover
       placement="bottom end"
       offset={0}
-      style={{ ...styles.tooltip, padding: 0, ...style }}
+      className={`${css({
+        ...styles.tooltip,
+        ...styles.lightScrollbar,
```
[the remark this answers is not in the document]
Comment (MatissJanis) [reply packages/desktop-client/src/components/common/Popover.tsx:21]: Yea, it did. But I think that was largely an unnoticed minor bug.

There are only 2x places where tooltips have scrollable content: autocomplete (where it makes sense to have `darkScrollbar` as the autocomplete is dark itself) and notes (where it makes more sense to have it light IMO).

**Comment by MatissJanis** [comment]
@joel-jeremy @youngcw 

What's the verdict here from you? Shall we proceed with the refactors to react-aria (which has some scrollbar changes) or would you prefer to roll-back the refactors already done (mostly just notes popovers)?

AFAIK the only blocker at this point is the scrollbar width change for MS/Linux devices (colors should be correct now?). If we decide to proceed with the migration - I'd create a separate issue to address the scrollbar widths and I'd hope on of yous could pick it up (as again: I do not have a way to reproduce the issue on Mac).

Let me know your thoughts.

**Comment by joel-jeremy** [comment]
@MatissJanis @youngcw Pushed an update for the scrollbar. Can you check?

**Comment by youngcw** [comment]
The scroll bars are fixed for me now.

I am also seeing the overlap in the bulk edit modal

**Comment by MatissJanis** [comment]
Thanks @joel-jeremy for figuring out the solution there!

Bulk edit modal should now be fixed.

**Comment by youngcw** [comment]
Looks good to me

**Review by joel-jeremy — APPROVED** [review]
LGTM!
